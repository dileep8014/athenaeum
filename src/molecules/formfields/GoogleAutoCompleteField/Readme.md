GoogleAutoCompleteField Example:

**You must provide either a `googlePlacesLink` or `googlePlacesAPIKey` prop for this example to work**

```jsx
  <GoogleAutoCompleteField
    input={{
      name: 'street',
    }}
    autocompleteAddressFields={(data) => { alert(JSON.stringify(data)) }}
  />
```
