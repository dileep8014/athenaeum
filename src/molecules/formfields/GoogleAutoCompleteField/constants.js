/* eslint-disable camelcase */
export const GOOGLE_ADDRESS_VALUES_MAP = {
  street_number: 'number',
  route: 'street',
  locality: 'city',
  sublocality_level_1: 'city',
  neighborhood: 'neighborhood',
  administrative_area_level_1: 'state',
  administrative_area_level_2: 'county',
  country: 'country',
  postal_code: 'zip'
};
