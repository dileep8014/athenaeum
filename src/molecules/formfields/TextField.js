import React from 'react';
import styles from './formfields.module.scss';
import classnames from 'classnames';

function TextField ( {
  className,
  placeholder,
  label
}) {

  return (
    <div className={ classnames(className, styles.textfield) }>
      <label
        className={ styles.label }
      >{ label }</label>
      <input
        className={ styles.input }
        type='text'
        placeholder={ placeholder } />
    </div>
  )
}

TextField.propTypes = {
  /**
   * Label is optional. If not provided, component will reorganize accordingly.
   */
  label: React.PropTypes.string,
  /**
   * Will append new classname to classSet
   */
  className: React.PropTypes.string,
  placeholder: React.PropTypes.string,
}

TextField.defaultProps = {
  placeholder: 'Placeholder'
}

export default TextField;

