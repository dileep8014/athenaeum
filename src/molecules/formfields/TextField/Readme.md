First name Field:

```example
    class TextFieldExample extends React.Component {
      constructor() {
        super();

        this.state = {
          value: ''
        };

        this.onChange = this.onChange.bind(this);
      }

      onChange(event) {
        this.setState({ value: event.target.value });
      }

      render() {
        return (
          <TextField
            label='First Name'
            placeholder='Enter your first name here'
            input={{
              value: this.state.value,
              onChange: this.onChange
            }}
          />
        )
      }
    }

    <TextFieldExample />
```

Text field with error:

```example
  <TextField
    label='First Name'
    placeholder='Enter your first name'
    input={{
      value: state.value,
      onChange: (event) => setState({ value: event.target.value })
    }}
    meta={{
      error: 'Maecenas ullamcorper turpis quis elit volutpat',
      touched: true
    }}
  />
```

Text field - phone number example:

```example
  <TextField
    label='Phone Number'
    placeholder='In case of emergency'
    input={{
      value: state.value,
      onChange: event => setState({ value: event.target.value })
    }}
    mask='999-999-9999'
    maskChar='_'
  />
```

Text field - secure:

```example
  <TextField
    label='Phone Number'
    placeholder='In case of emergency'
    input={{
      value: state.value,
      onChange: event => setState({ value: event.target.value })
    }}
    mask='999-999-9999'
    maskChar='_'
    secure
  />
```
