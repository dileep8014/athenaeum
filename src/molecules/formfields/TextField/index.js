/* eslint-disable complexity */
import React from 'react';
import PropTypes from 'prop-types';
import InputMask from 'react-input-mask';
import classnames from 'classnames';
import omit from 'lodash/omit';

import Icon from 'atoms/Icon';
import Text from 'atoms/Text';
import ErrorMessage from 'atoms/ErrorMessage';

import { renderTooltip } from 'utils/fieldUtils';

import styles from './text_field.module.scss';

class TextField extends React.Component {
  render() {
    const {
      className,
      htmlFor,
      input,
      label,
      mask,
      maskChar,
      meta,
      noBaseStyle,
      placeholder,
      secure,
      id,
      tooltip,
      inputTooltip,
      type,
      maskVal,
      prefix,
      postfix,
      activeValidation,
      subLabel,
      swiftype,
      fieldRef,
    } = this.props;

    const classes = [
      !noBaseStyle && styles['text-field'],
      meta && meta.active && !noBaseStyle && styles['focused'],
      meta && meta.touched && meta.error && !meta.active && styles['hasError'],
      className,
    ];

    const inputClasses = [
      styles['input-wrapper'],
      prefix && styles['prefix'],
      postfix && styles['postfix'],
      swiftype && styles['swiftype-prefix'],
    ];

    let message;

    if (activeValidation) {
      message = meta && (meta.dirty || meta.touched) && (meta.error || meta.warning);
    } else {
      message = meta && meta.touched && !meta.active && (meta.error || meta.warning);
    }

    return (
      <div ref={fieldRef && fieldRef}>
        <div className={classnames(...classes)}>
          { label && (
            <div className={classnames(styles['header'])}>
              <div className={styles['label-wrapper']}>
                {/* eslint-disable-next-line jsx-a11y/label-has-for */}
                <label className={styles['label']} htmlFor={id || htmlFor}>{ label }</label>
                { secure && <Icon className={styles['icon-lock']} icon='lock' /> }
                { tooltip && renderTooltip(tooltip, styles['tooltip'], styles['tooltip-icon']) }
              </div>
              <Text size={10} font='b'>
                { subLabel }
              </Text>
            </div>
          )}
          <span
            data-prefix={prefix}
            data-postfix={postfix}
            className={classnames(...inputClasses)}
          >
            { swiftype && <Icon className={styles['icon-search']} icon='searchRebrand' /> }
            { mask
              ? (
                <InputMask
                  className={styles['input']}
                  type={type}
                  placeholder={placeholder}
                  mask={mask}
                  maskChar={maskChar}
                  id={id}
                  {...input}
                />
              )
              : (
                <input
                  className={classnames(
                    styles['input'],
                    swiftype && 'swiftype-search'
                  )}
                  type={type}
                  placeholder={placeholder}
                  id={id}
                  value={maskVal(input.value)}
                  {...omit(input, [ 'value' ])}
                />
              )}
            {inputTooltip && renderTooltip(inputTooltip, styles['input-tooltip'], styles['tooltip-icon'])}
          </span>
        </div>

        <ErrorMessage
          condition={!!message}
          message={message}
          type={meta && meta.warning ? 'warning' : 'error'}
        />
      </div>
    );
  }
}

TextField.propTypes = {

  /**
   * set to true to allow validation errors/warnings to occur `onChange`
   */
  activeValidation: PropTypes.bool,

  /**
   * Label is optional. If not provided, component will reorganize accordingly.
   */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),

  /**
   * `for` prop on label.
   * Please use the `id` prop instead
   */
  htmlFor: PropTypes.string,

  /**
   * Will append new classname to classSet
   */
  className: PropTypes.string,

  /**
   * placeholder text for text field
   */
  placeholder: PropTypes.string,

  /**
   * object with all necessary props for input
   */
  input: PropTypes.object,

  /**
   * The props under the meta key are metadata about the state of this field that `redux-form` tracks.
   */
  meta: PropTypes.object,

  /**
   * Passing `noBaseStyle=true` will omit the base class style
   * Fields may at times be wrapped up as a collection. In this case, the wrapper would control the styling for those fields as a group.
   */
  noBaseStyle: PropTypes.bool,

  /**
   * Mask the input
   */
  mask: PropTypes.string,

  /**
   * Specific the maskChar (must have `mask` set for this to be applied)
   */
  maskChar: PropTypes.string,

  /**
   * Function to mask the actual input value
   *
   * Provides for more specific customization over the `mask` prop
   */
  maskVal: PropTypes.func,

  /**
   * Adds a lock to the label.
   */
  secure: PropTypes.bool,

  /**
   * id added to the `input` node and used for the `for` HTML attribute on the associated `label`.
   * This prop is required to ensure the `label` and `input` follow best HTML5 accessibility practices as well as for testing purposes
   */
  id: PropTypes.string.isRequired,

  /**
   * either a handler for clicking the tooltip, or text to go in the tooltip for the label
   */
  tooltip: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
    PropTypes.object
  ]),

  /**
   * either a handler for clicking the tooltip, or text to go in the tooltip for the input
   */
  inputTooltip: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
  ]),

  /**
   * HTML type for input node; when used with `mask` it always defaults to 'text'
   */
  type: PropTypes.oneOf([
    'text',
    'password',
    'tel',
  ]),

  /**
   * provides a prefix for the input
   */
  prefix: PropTypes.oneOf([
    '$',
  ]),

  /**
   * provides a postfix for the inpit
   */
  postfix: PropTypes.string,

  /**
   * Options text displayed directly underneath label
   */
  subLabel: PropTypes.string,

  /**
   * Enables swiftype search
   */
  swiftype: PropTypes.bool,

  /**
   * Applies a React ref to the wrapping node for this field
   */
  fieldRef: PropTypes.func,

  errorMessage: PropTypes.bool
};

TextField.defaultProps = {
  errorMessage: false,
  maskVal: (val) => val,
  placeholder: 'Placeholder',
  type: 'text',
};

export default TextField;
