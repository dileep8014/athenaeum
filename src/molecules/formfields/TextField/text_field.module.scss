@import '../shared/formfields.module.scss';

.text-field {
  @extend %form-field;
}

.header {
  display: flex;
  align-items: center;
  padding-right: ru(.5);

  border-bottom: $border;
}

.icon-lock {
  width: ru(.75);
  height: ru(.75);
  margin-left: auto;
}

.label {
  @extend %label;
}

.input {
  @extend %field;
}

input[type='text'] {
  outline: none;
}

/* --  state:focus  -- */
.focused {
  @extend %focused;
}

/* --  state:error  -- */
.hasError {
  @extend %hasError;
}
