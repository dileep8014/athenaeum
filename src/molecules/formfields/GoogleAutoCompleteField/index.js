/* eslint-disable no-undef */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import pick from 'lodash/pick';
import reduce from 'lodash/reduce';
import keys from 'lodash/keys';
import assign from 'lodash/assign';
import TextField from 'molecules/formfields/TextField';

import loadJS from './util/loadJS';
import {
  GOOGLE_ADDRESS_VALUES_MAP,
} from './constants';

class GoogleAutoCompleteField extends Component {
  state = {
    autocomplete: null,
  }

  componentDidMount() {
    const {
      googlePlacesLink,
      googlePlacesAPIKey
    } = this.props;

    window.initMap = this.initMap;

    if (googlePlacesLink || googlePlacesAPIKey) {
      loadJS({
        link: googlePlacesLink,
        key: googlePlacesAPIKey,
      });
    }
  }

  initMap = () => {
    const { id } = this.props;

    const autocomplete = new google.maps.places.Autocomplete(
      document.getElementById(`${id}`),
      { types: [ 'geocode' ] }
    );

    this.setState({ autocomplete });

    autocomplete.addListener('place_changed', () => {
      this.updateAddress();
    });
  }

  updateAddress = () => {
    const place = this.state.autocomplete.getPlace();
    const addressData = this.formatAddressInfo(place.address_components);

    if (!addressData.number) {
      const { value } = this.props.input;
      const streetNumberMatch = value && value.match(/^(\d+)/);

      if (streetNumberMatch && streetNumberMatch[0]) {
        const number = parseInt(streetNumberMatch[0], 10);

        addressData.number = { shortName: number, longName: number };
      }
    }

    const data = this.pickData(addressData);

    this.props.autocompleteAddressFields(data);
  }

  pickData = (addressData) => {
    const { addressValuesToPick } = this.props;
    const pickedData = pick(addressData, keys(addressValuesToPick));

    // Grab the short or long name from the addressData based on props.addressValuesToPick
    return reduce(pickedData, (res, names, addressValue) => assign({}, res, {
      [addressValue]: names[addressValuesToPick[addressValue]]
    }), {});
  }

  formatAddressInfo(addressInfo) {
    // Return an object where the laymen term value references both short_name and long_name
    // Ex:
    // info = {
    //   short_name: 'Broadway',
    //   long_name: 'Broadway,
    //   types: ['route'],
    // }
    //
    // becomes
    //
    // street: {
    //  shortName: 'Broadway',
    //  longName: 'Broadway,
    // }

    return addressInfo.reduce((res, info) => {
      const addressType = info.types[0];
      const formattedAddressValue = GOOGLE_ADDRESS_VALUES_MAP[addressType];

      if (formattedAddressValue) {
        return assign({}, res, {
          [formattedAddressValue]: {
            shortName: info.short_name,
            longName: info.long_name,
          }
        });
      }

      return res;
    }, {});
  }

  render() {
    const { id } = this.props;

    return (
      <TextField
        {...this.props}
        id={id}
      />
    );
  }
}

GoogleAutoCompleteField.propTypes = {
  /**
   * Values to be passed down to the TextField rendered by this component.
   * If using `redux-form`, this prop will come from the `Field` wrapper component.
   * `name` is required in order to build the Google autocomplete data properly.
   */
  input: PropTypes.shape({
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
  }).isRequired,

  /**
   * Function that will be fired when a user selects an address from the autocomplete drop down.
   * It will be passed a `data` object containing the values picked from the Google autocomplete response
   */
  autocompleteAddressFields: PropTypes.func.isRequired,

  /**
   * `id` prop onto which Google will hook the autocomplete functionality
   */
  id: PropTypes.string.isRequired,

  /**
   * A map of values to pick from the Google autocomplete response.
   * Available values to pick are:
   *
   * ```
   *  number
   *  street
   *  city
   *  neighborhood
   *  state
   *  county
   *  country
   *  zip
   * ```
   *
   * Available formatted values are: `shortName` and `longName`.
   *
   * You can construct this map as follows to receive the short name for number and long name for city:
   *
   * ```
   * {
      number: 'shortName',
      city: 'longName',
   * }
   * ```
   */
  addressValuesToPick: PropTypes.object,

  /**
   * Google Places authorized API link.
   * Providing this link will automatically load the necessary JS to enable the autocomplete functionality.
   * It is recommended that you as the consumer handle loading the Google JS to prevent multiple loads if you use this component more than once.
   */
  googlePlacesLink: PropTypes.string,

  /**
   * Google Places authorized API Key.
   * Providing this key will automatically load the necessary JS to enable the autocomplete functionality.
   * It is recommended that you as the consumer handle loading the Google JS to prevent multiple loads if you use this component more than once.
   */
  googlePlacesAPIKey: PropTypes.string,
};

GoogleAutoCompleteField.defaultProps = {
  id: 'autocomplete',
  addressValuesToPick: {
    number: 'shortName',
    street: 'longName',
    city: 'longName',
    state: 'longName',
    zip: 'shortName',
  },
};

export default GoogleAutoCompleteField;
