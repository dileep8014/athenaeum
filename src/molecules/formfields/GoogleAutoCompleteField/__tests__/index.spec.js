import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import TextField from 'molecules/formfields/TextField';

import GoogleAutoCompleteField from '../';

describe('<GoogleAutoCompleteField />', () => {
  let props;
  let addressValues;

  beforeEach(() => {
    props = {
      input: {
        name: 'streetAddress',
      },
      label: 'Street address',
      placeholder: 'Enter your street address',
      autocompleteAddressFields: jest.fn(),
    };

    addressValues = [
      {
        types: [
          'street_number'
        ],
        short_name: '1',
        long_name: '1',
      },
      {
        types: [
          'route'
        ],
        long_name: 'Broadway',
        short_name: 'Broadway',
      },
      {
        types: [
          'sublocality_level_1'
        ],
        long_name: 'New York',
        short_name: 'New York',
      },
      {
        types: [
          'administrative_area_level_1'
        ],
        short_name: 'NY',
        long_name: 'New York',
      },
      {
        types: [
          'postal_code'
        ],
        short_name: '10075',
        long_name: '10075',
      },
    ];
  });

  it('renders correctly', () => {
    const wrapper = shallow(<GoogleAutoCompleteField {...props} />);

    const actual = renderer.create(wrapper).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('returns a single TextField with appropriate props', () => {
    const wrapper = shallow(<GoogleAutoCompleteField {...props} />);

    const field = wrapper.find(TextField);

    expect(field.length).toEqual(1);
    expect(field.props().name).toEqual(props.name);
    expect(field.props().label).toEqual('Street address');
    expect(field.props().placeholder).toEqual('Enter your street address');
    expect(field.props().id).toEqual('autocomplete');
  });

  describe('updateAddress', () => {
    it('calls autocomplete.getPlace and props.autocompleteAddressFields', () => {
      const expectedData = {
        number: '1',
        street: 'Broadway',
        city: 'New York',
        state: 'New York',
        zip: '10075',
      };

      const state = {
        autocomplete: {
          getPlace: jest.fn(() => {
            return {
              address_components: addressValues,
            };
          })
        }
      };

      const wrapper = shallow(<GoogleAutoCompleteField {...props} />);

      wrapper.setState(state);

      wrapper.instance().updateAddress();

      expect(state.autocomplete.getPlace.mock.calls.length).toBe(1);
      expect(props.autocompleteAddressFields.mock.calls.length).toBe(1);
      expect(props.autocompleteAddressFields.mock.calls[0][0]).toEqual(expectedData);
    });
  });

  describe('formatAddressInfo', () => {
    it('returns the correctly formatted data', () => {
      const expectedData = {
        number: {
          shortName: '1',
          longName: '1',
        },
        street: {
          shortName: 'Broadway',
          longName: 'Broadway',
        },
        city: {
          shortName: 'New York',
          longName: 'New York',
        },
        state: {
          shortName: 'NY',
          longName: 'New York',
        },
        zip: {
          shortName: '10075',
          longName: '10075',
        },
      };

      const wrapper = shallow(<GoogleAutoCompleteField {...props} />);

      expect(wrapper.instance().formatAddressInfo(addressValues)).toEqual(expectedData);
    });
  });

  describe('pickData', () => {
    it('returns the correctly picked data', () => {
      const expectedData = {
        number: '1',
        street: 'Broadway',
        city: 'New York',
        state: 'New York',
        zip: '10075',
      };

      const wrapper = shallow(<GoogleAutoCompleteField {...props} />);

      const data = wrapper.instance().formatAddressInfo(addressValues);

      expect(wrapper.instance().pickData(data)).toEqual(expectedData);
    });
  });
});
