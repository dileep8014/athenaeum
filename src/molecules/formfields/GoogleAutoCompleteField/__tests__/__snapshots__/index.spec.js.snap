// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoogleAutoCompleteField /> renders correctly 1`] = `
<div>
  <div
    className="text-field"
  >
    <div
      className="header"
    >
      <div
        className="label-wrapper"
      >
        <label
          className="label"
          htmlFor={undefined}
        >
          Street address
        </label>
      </div>
    </div>
    <span
      className="input-wrapper"
      data-postfix={undefined}
      data-prefix={undefined}
    >
      <input
        className="input"
        id="autocomplete"
        name="streetAddress"
        placeholder="Enter your street address"
        type="text"
        value={undefined}
      />
    </span>
  </div>
  <span />
</div>
`;
